pyirf/interpolation.py

```diff
@@ -2,22 +2,238 @@
+
+        # Predefine selection mask
+        selection = np.ones(len(cdf), dtype=bool)
+
+        # Keep only the first of subsequently matching values
+        selection[1:] = cdf[1:] != cdf[:-1]
+
+        # Keep only the last 0 and first 1 entry
+        selection[:last_0] = False
+        selection[last_0] = True
+        selection[first_1 + 1 :] = False
+        selection[first_1] = True
+
+        # create ppf values from selected bins
+        return interp1d(
+            cdf[selection],
+            bin_mids[selection],
+            bounds_error=False,
+            fill_value="extrapolate",
+        )(quantiles)
+
+    # create ppf values from cdf samples via interpolation of the cdfs
+    # return nan for empty pdfs
+    ppfs = np.apply_along_axis(
+        lambda cdf: cropped_interp(cdf)
+        if np.sum(cdf) > 0
+        else np.full_like(quantiles, np.nan),
+        -1,
+        cdfs,
+    )
+    # nD interpolation of ppf values
+    return griddata(m, ppfs, mprime)
+
+
+def reconstruct_pdf_values(quantiles, ppfs, edges):
+    """
+    Reconstruct pdf from ppf and evaluate at desired points.
+
+    Parameters
+    ----------
+    quantiles: numpy.ndarray, shape=(L)
+        L quantiles for which the ppf_values are known
+
+    ppfs: numpy.ndarray, shape=(1,...,L)
+        Corresponding ppf-values for all quantiles
+
+    edges: numpy.ndarray, shape=(M+1)
+        Binning of the desired binned pdf
+
+    Returns
+    -------
+    pdf_values: numpy.ndarray, shape=(1,...,M)
+        Recomputed, binned pdf
+    """
+    # recalculate pdf values through numerical differentiation
+    pdf_interpolant = np.nan_to_num(np.diff(quantiles) / np.diff(ppfs, axis=-1))
+
+    # Unconventional solution to make this usable with np.apply_along_axis for readability
+    # The ppf bin-mids are computed since the pdf-values are derived through derivation
+    # from the ppf-values
+    xyconcat = np.concatenate(
+        (ppfs[..., :-1] + np.diff(ppfs) / 2, pdf_interpolant), axis=-1
+    )
+
+    # Interpolate pdf samples and evaluate at bin edges, weight with the bin_width to estimate
+    # correct bin height via the midpoint rule formulation of the trapezoidal rule
+    pdf_values = np.apply_along_axis(
+        lambda xy: np.diff(edges)
+        * np.nan_to_num(
+            interp1d(
+                xy[: int(len(xy) / 2)],
+                xy[int(len(xy) / 2) :],
+                bounds_error=False,
+                fill_value=(0, 0),
+            )(edges[:-1])
+        ),
+        -1,
+        xyconcat,
+    )
+
+    return pdf_values
+
+
+def norm_pdf(pdf_values):
+    """
+    Normalize binned_pdf to a sum of 1
+    """
+    norm = np.sum(pdf_values, axis=-1)
+
+    # Ignore errors from rows without errors as they are set to 0 through nan_to_num
+    with np.errstate(invalid="ignore"):
+        normed_pdf_values = pdf_values / norm[..., np.newaxis]
+    return np.nan_to_num(normed_pdf_values)
```

Comment:
I think the solution before was  a tad more elegant:
```
normed_pdf_values = np.divide(pdf_values, norm, out=np.zeros_like(pdf_values), where=norm != 0)```

Reply:
This would also have the advantage of preserving real nans that did not come from dividing through a 0 norm here but were there before

Reply:
This results in errors as numpy seems to be unable to broadcast this correctly